crates/nu-parser/src/parse_literals.rs

```diff
@@ -0,0 +1,573 @@
+        Ok(unsafe { std::str::from_utf8_unchecked(maybe_ascii) })
+    } else {
+        Err(AsciiError)
+    }
+}
+
+pub fn parse_int(working_set: &mut StateWorkingSet, span: Span) -> Expression {
+    let token = working_set.get_span_contents(span);
+
+    fn parse_int_with_base(token: &[u8], span: Span, radix: u32) -> Result<Expression, ParseError> {
+        let token = strip_underscores(token);
+        if let Ok(token) = parse_ascii_as_str(&token) {
+            if let Ok(num) = i64::from_str_radix(token, radix) {
+                return Ok(Expression {
+                    expr: Expr::Int(num),
+                    span,
+                    ty: Type::Int,
+                    custom_completion: None,
+                });
+            }
+        }
+        Err(ParseError::InvalidLiteral(
+            format!("invalid digits for radix {}", radix),
+            "int".into(),
+            span,
+        ))
+    }
+
+    if token.is_empty() {
+        working_set.error(ParseError::Expected("int", span));
+        return garbage(span);
+    }
+
+    let res = if let Some(num) = token.strip_prefix(b"0b") {
+        parse_int_with_base(num, span, 2)
+    } else if let Some(num) = token.strip_prefix(b"0o") {
+        parse_int_with_base(num, span, 8)
+    } else if let Some(num) = token.strip_prefix(b"0x") {
+        parse_int_with_base(num, span, 16)
+    } else {
+        let token = strip_underscores(token);
+        parse_ascii_as_str(&token)
+            .map_err(|_| ParseError::Expected("int", span))
+            .and_then(|s| {
+                s.parse::<i64>()
+                    .map(|num| Expression {
+                        expr: Expr::Int(num),
+                        span,
+                        ty: Type::Int,
+                        custom_completion: None,
+                    })
+                    .map_err(|_| ParseError::Expected("int", span))
+            })
+    };
+
+    match res {
+        Ok(expr) => expr,
+        Err(err) => {
+            working_set.error(err);
+            garbage(span)
+        }
+    }
+}
+
+pub fn parse_float(working_set: &mut StateWorkingSet, span: Span) -> Expression {
+    let token = working_set.get_span_contents(span);
+    // TODO: we should probably disallow underscores in the special IEEE754 strings
+    // e.g. `N_a_N` should not parse as a float
+    let token = strip_underscores(token);
+
+    if let Ok(token) = parse_ascii_as_str(&token) {
+        if let Ok(x) = token.parse::<f64>() {
+            return Expression {
+                expr: Expr::Float(x),
+                span,
+                ty: Type::Float,
+                custom_completion: None,
+            };
+        }
+    }
+    working_set.error(ParseError::Expected("float", span));
+    garbage(span)
+}
+
+pub fn parse_number(working_set: &mut StateWorkingSet, span: Span) -> Expression {
+    let starting_error_count = working_set.parse_errors.len();
+
+    let result = parse_int(working_set, span);
```

Comment:
it might be good to be able to signal the error without having to check `parse_errors`, as creating the error would definitely lead to some allocs. but not critical

maybe this can be done by checking if it looks like an int or float first, rather than fallback?

Reply:
That is a pervasive pattern accross the parser to always push the error and then remove if necessary as part of the way how we allow multiple IDE errors to appear. So `working_set.parse_errors` is likely properly sized and hot.

From the way the functions look/have to be handled its a bit unwieldy and potentially error prone.

Reply:
Probably a larger effort but we could probably move a bunch of parsing steps to use specific string-free error types/enums that only later get turned into a `ParseError` with `String` allocations if we an alternative mechanism to the one big error stack.